Automated cherry pick of #115620: client-go/cache: fix missing delete event on replace (+ #116623)

odinuge opened the pull request:
Cherry pick of #115620 on release-1.24.

#115620: client-go/cache: fix missing delete event on replace

For details on the cherry pick process, see the [cherry pick requests](https://git.k8s.io/community/contributors/devel/sig-release/cherry-picks.md) page.

```release-note
Fix missing delete events on informer re-lists to ensure all delete events are correctly emitted and using the latest known object state, so that all event handlers and stores always reflect the actual apiserver state as best as possible
```

odinuge committed:
client-go/cache: fix missing delete event on replace

This fixes a race condition when a "short lived" object
is created and the create event is still present on the queue
when a relist replaces the state. Previously that would lead in the
object being leaked.

The way this could happen is roughly;

1. new Object is added O, agent gets CREATED event for it
2. watch is terminated, and the agent runs a new list, L
3. CREATE event for O is still on the queue to be processed.
4. informer replaces the old data in store with L, and O is not in L
  - Since O is not in the store, and not in the list L, no DELETED event
    is queued
5. CREATE event for O is still on the queue to be processed.
6. CREATE event for O is processed
7. O is <leaked>; its present in the cache but not in k8s.

With this patch, on step 4. above it would create a DELETED event
ensuring that the object will be removed.

Signed-off-by: Odin Ugedal <ougedal@palantir.com>
Signed-off-by: Odin Ugedal <odin@uged.al>

odinuge committed:
client-go/cache: fix missing delete event on replace without knownObjects

This fixes an issue where a relist could result in a DELETED delta
with an object wrapped in a DeletedFinalStateUnknown object; and then on
the next relist, it would wrap that object inside another
DeletedFinalStateUnknown, leaving the user with a "double" layer
of DeletedFinalStateUnknown's.

Signed-off-by: Odin Ugedal <ougedal@palantir.com>
Signed-off-by: Odin Ugedal <odin@uged.al>

odinuge committed:
client-go/cache: merge ReplaceMakesDeletionsForObjectsInQueue tests

Signed-off-by: Odin Ugedal <ougedal@palantir.com>
Signed-off-by: Odin Ugedal <odin@uged.al>

odinuge committed:
client-go/cache: rewrite Replace to check queue first

This is useful to both reduce the code complexity, and to ensure clients
get the "newest" version of an object known when its deleted. This is
all best-effort, but for clients it makes more sense giving them the
newest object they observed rather than an old one.

This is especially useful when an object is recreated. eg.

Object A with key K is in the KnownObjects store;
- DELETE delta for A is queued with key K
- CREATE delta for B is queued with key K
- Replace without any object with key K in it.

In this situation its better to create a DELETE delta with
DeletedFinalStateUnknown with B (with this patch), than it is to give
the client an DeletedFinalStateUnknown with A (without this patch).

Signed-off-by: Odin Ugedal <ougedal@palantir.com>
Signed-off-by: Odin Ugedal <odin@uged.al>

odinuge committed:
client-go/cache: update Replace comment to be more clear

Since the behavior is now changed, and the old behavior leaked objects,
this adds a new comment about how Replace works.

Signed-off-by: Odin Ugedal <ougedal@palantir.com>
Signed-off-by: Odin Ugedal <odin@uged.al>

wojtek-t commented:
/lgtm
/approve

@kubernetes/release-managers - for cherrypick approval

wojtek-t commented:
/kind bug
/priority important-soon

cici37 commented:
/triage accepted

lavalamp commented:
/lgtm
/approve

odinuge commented:
Adding a hold while we discuss https://github.com/kubernetes/kubernetes/issues/115658#issuecomment-1450614656.

/hold

Verolop commented:
@lavalamp per @odinuge last comment on https://github.com/kubernetes/kubernetes/issues/115658#issuecomment-1463661460, I'm happy to add this as a cherry pick for the upcoming patches. Can you please remove the hold tag? thanks!

lavalamp committed:
Change where transformers are called.

odinuge: sorted out some function signature changes during
cherry-picking that caused conflicts.

(cherry picked from commit e76dff38cf74c3c8ad9ed4d3bc6e3641d9b64565)
Signed-off-by: Odin Ugedal <odin@uged.al>

lavalamp commented:
Please add https://github.com/kubernetes/kubernetes/pull/116623 to this cherry pick and then it should be good to go.

(sorry for copy-pasting this everywhere :) )

odinuge commented:
/hold cancel

cc @kubernetes/release-managers for approval

odinuge commented:
> Shouldn't the release note be added together as in original PR?

Its been a while, but from the docs it looks like its supposed to be applied automatically;
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-release/cherry-picks.md
> The same release note requirements apply as normal pull requests, except the release note stanza will auto-populate from the master branch pull request from which the cherry pick originated.

I'm happy to add it if ^ is not the case

odinuge commented:
Well, added it to all of them now to be sure it works as expected. Thanks

xmudrii reviewed (approved):
For RelEng:
/lgtm
/approve
